### ISSUE TITLE
minipay - minipay

### ISSUE THREAD

**Issue by mihbor** [minima-global/innovation-challenge#15]
**Before you start, please update the issue title:**

Example: *Minima DEX - The A Team*

## :pencil2: Project & Team Details

1. MiniDapp/Project Name [required, as used in your Google Form]: *Provide the name of your MiniDapp/Project* - minipay

2. MiniDapp/Project Description [optional, if changed from your application in Google Form]: *Please provide a brief description of your MiniDapp/Project*

3. Team Name [optional, if changed from your application in Google Form]: *Provide your team name*

4. Team Members [optional, if changed from your application in Google Form]: *Please list names of all team members*

## :one: Phase 1 Submission 

For Phase 1, please upload the following 
(paste the link to where you uploaded e.g. other cloud provider or GitHub (if more than 100MB) or Attach files by dragging & dropping, selecting or pasting them)

5. A whitepaper or a pitch video with a description of the idea you have that includes: your idea, how it meets the criteria to be considered for the competition, the benefits it will bring to people and how the unique features of the Minima network can make a real difference/give an advantage to your solution.

[minipay whitepaper.pdf](https://github.com/minima-global/innovation-challenge/files/9271201/minipay.whitepaper.pdf)

6. Upload Presentation: *Upload any relevant presentations/slide decks you have created*

https://docs.google.com/presentation/d/1qnujdCyc8KMqzUWc7fxXH6ziqxmjAs7mkx_ek62neKI/edit?usp=sharing

[minipay.pdf](https://github.com/minima-global/innovation-challenge/files/9271183/minipay.pdf)

## :two: Phase 2 Submission 

For Phase 2, please update your submission, providing the following:

7. Code Repository Link: *Provide a link to the Github URL (or other) for your MiniDapp/Project*

8. Upload MiniDapp: *If your project includes a MiniDapp, please upload your MiniDapp - this should be a .zip file using the **.mds.zip** extension*

9. Upload Demo: *Upload any demonstration videos you have created (optional)*

##  :round_pushpin: Other Comments
*Optional - add any other comments regarding your submission*

## :heavy_check_mark: Checklist

**Before you post the issue**:
- [x] Make sure you have entered using this [Google Form](https://docs.google.com/forms/d/e/1FAIpQLSdW-r1iTN2JUxaLhKZxLj4FgRnIF6yZWAXB30hj4c-vwaNuPw/viewform)
- [x] You have followed the issue title format
- [x] You have provided all the information correctly and uploaded all required material for the Phase you are submitting for
- [x] You have read and agree with the terms in the [Code Of Conduct]()

**Comment by martapiekarska** [comment]
Hi Team! Can I please ask you to confirm that you are working on your final submission - the final deadline is 12th of September. Please try taking the continuous submission approach rather than send everything at once. This will help judges and show the journey you’re taking and the thought process.

**Comment by mihbor** [comment]
Hi @martapiekarska ,

I can report good progress. 

The project is made of 2 subprojects, each in it's own repo for now. One for the browser-based minidapp, one for the native Android app.

Of the original roadmap presented in stage 1 submission:
1) pay to an address by scanning a QR code or tapping NFC (another
mobile device or a pre-programmed tag).
2) request a payment to your address by presenting a QR code or your
phone's NFC device or using the NFC in your phone to program an NFC
tag sticker (for merchants).
3) establish a Layer 2 payment channel using QR codes or NFC and
Maxima for P2P messaging.
4) pay and request payments over an existing L2 channel.
5) pay and request payments using NFC over existing L2 channel while
offline.
6) payment channel route discovery via intermediaries

the progress update is:
1 and 2) - Pay and Request to Pay by QR code is working in the minidapp (although the QR scanning video component doesn't working inside the MDS hub, have to click "Open in browser") and the Pay and Request to Pay by NFC (contactless) is working in the Android app. The minidapp also has a link to download the Android .apk (again, doesn't work in MDS hub, but works in browser). Once installed, you can pre-fill the Request to Pay (address, amount, token) in the minidapp and click a "deep link" that opens the Android app with the same data (address, amount, token) filled in and stats emitting via NFC.

3 and 4) payment channel works in the minidapp (only) as of today! :) you can create a channel, update balances (send coins) off-chain and settle the channel on-chain. The current limitations are:
* when opening the channel, only one party funds it
* in the off-chain update phase you can send coins but you cannot request coins be sent to you (yet). Technically, the feature to request funds is already there but not processed on the receiving side, yet.
* channel communication is via firebase, which is a point of centralisation - to be replaced by maxima later
These will be tackled later.
The bigger limitation and the next priority is that there is no persistence of the channel state yet. The idea is to store it using MDS sql commands and query it to display channel listing and allow to continue across browser restarts. This will also allow the Android app to read channel state and partake in the same channels as the browser minidapp. 

I expect implementing persistence will be fiddly and will take some time but can be achieved in the timelines. It is a prerequisite for 5) offline payments. Not sure if time will allow to do this but it would be a great demonstrator, IMO. To be able to transaction on a pre-existing channel between 2 parties even if without internet. 

6) multi-hop payments is ambitious and certainly out of scope of the PoC, but that is the route to commercialization, as you can charge fees for using your channel and provide payment route discovery services for a fee.
Another commercial opportunity is a service that checks (again for a fee), that your counterparties are not trying to settle the channel with stale transactions while you're not looking.

I'll try to add some screenshots of working functionality over the weekend.

**Comment by mihbor** [comment]
The minidapp in the MDS Hub:
<img src="https://user-images.githubusercontent.com/6325352/187075341-85dcc1d6-b022-4f00-bffd-168d25c75f1f.jpg" width=270 height=600/>

Request view renders QR code (in MDS Hub):
<img src="https://user-images.githubusercontent.com/6325352/187075403-c605b1e8-962f-4cff-9a9c-2a4e84f8dc64.jpg" width=270 height=600/>

Send view should show camera component but doesn't work in MDS hub (works in Chrome, via Open in Browser):
<img src="https://user-images.githubusercontent.com/6325352/187075451-f0cc17ef-5a10-4341-83be-ccc92d28bf7f.jpg" width=270 height=600/>

Request view renders QR code (in Chrome):
<img src="https://user-images.githubusercontent.com/6325352/187075576-aae2061d-2c24-4327-a0f6-05de0a2421d4.jpg" width=270 height=600/>

Clicking "Request on NFC" in Chrome (doesn't work in MDS Hub directly) brings up the Android app (if installed):
<img src="https://user-images.githubusercontent.com/6325352/187075592-8e2b0ec0-3afe-4dce-96e5-37afcec685c2.jpg" width=270 height=600/>

"Disabling reader mode" toast indicates the app starts emitting the data on NFC, rather than waiting to receive data from NFC.


More screenshots to follow laters

**Comment by mihbor** [comment]
Final progress update before submission, I think. 
Channel persistence now works :)
Updating channel over NFC also works, even offline, as intended :) :) 
I will post screenshots shortly and work on some kind of demo slideshow and video.

**Comment by mihbor** [comment]
Creating an eltoo payment channel works currently only in the browser minidapp and the QR scanner works only in a browser, not embedded in the MDS HUB.

1) Pressing "Request payment channel" gets 3 public keys from Minima (trigger key, update key and settlement key) and displays them along with a QR code generated from a string concatenation thereof, separated by semicolons. It also subscribes (using websocket) to firebase updates to a transaction with the ID made of those same keys. TODO: replace firebase with maxima, an item for the roadmap
![request-payment-channel](https://user-images.githubusercontent.com/6325352/189545994-d7aed676-8599-4094-b152-fa6f47e29e12.jpg)

2) Pressing "Fund payment channel" on another device gets 3 public keys from Minima and opens the QR scanner to scan the public keys of the counter party. It also provides text fields where they can be manually entered (useful when testing with both sides on a development machine).
![fund-payment-channel](https://user-images.githubusercontent.com/6325352/189546131-3a111aed-3119-4c69-bd67-45aaee452eee.jpg)

It's a current limitation of the app that only one of the parties funds the channel initially. Funding from both sides needs to be on the roadmap.

3) Once the QR code is successfully scanned, the scanner video component disappears and instead the following inputs appear:
* numeric text input to enter the amount to fund the channel (min = 0, TODO: max should be set to the sendable balance of selected token)
* token select drop-down containing the token name and sendable amount
* integer text input to enter the time-lock duration of the eltoo script in blocks (default 10, TODO: increase this)
* "Initiate!" button
![fund-channel-scanned](https://user-images.githubusercontent.com/6325352/189546403-f52737ef-d1ee-47a9-a3bb-0bba05cdc7d3.jpg)

4) Clicking "Initiate!" does a number of steps:
* prepares a funding transaction sending the selected amount to the multisig script address built using the trigger keys of both parties
* prepares and signs a trigger transaction spending the (floating) output of the funding transaction and sending it to the eltoo script address built using update and settlement keys of both parties and the chosen time-lock value
* prepares and signs an initial settlement transaction spending the (floating) output of the trigger transaction and sending to to the funding party (self) address
* stores all this info in the DB using MDS.sql along with status = "OFFERED", sequence number(the state used to count updates in the eltoo script) = 0 and auto incremented primary key.
* sends all 3 public keys, the exported trigger and settlement transactions along with the time-lock value to firebase, using the counterparty's keys as the ID.
* It also subscribes (using websocket) to firebase updates to a transaction with the ID made of its repective keys.
![fund-channel-sent](https://user-images.githubusercontent.com/6325352/189546773-a415207a-d261-497c-93e3-989b9d847cdb.jpg)

5) When the counter party receives a message from firebase with a new record snapshot, it does the following:
* hides the QR code
* uses the counterparty's public keys to construct the multisig and eltoo contracts
* imports, signs and exports the trigger and settlement transactions
* stores them along with status = "OFFERED", sequence number 0 and the other information in the DB using MDS.sql
* sends back the trigger and settlement transactions by saving them in firebase using the received counterparty's public keys as ID
* starts monitoring the balance of the multisig address upon every NEWBLOCK message from Minima
![request-channel-received](https://user-images.githubusercontent.com/6325352/189547027-a01929e3-e693-45d5-8ebd-5c56fcd21c87.jpg)

6. When the funding party receives a message from firebase with a new record snapshot, it does the following:
* imports the received trigger and settlement transactions which are now signed by both parties
* Updates its channel in the DB using MDS.sql command with the exported transactions
* Signs and posts the funding transaction
* Starts monitoring the balance of the multisig address upon every NEWBLOCK message from Minima

![fund-channel-posted](https://user-images.githubusercontent.com/6325352/189547206-10cd89cc-050c-402e-9f24-443d4fdef293.jpg)

7) When each party discovers new confirmed or unconfirmed balance at the multisig address, they update their channel status in the DB to "OPEN".

**Comment by mihbor** [comment]
Both the minidapp and the android app have a "Channel listing" button which when clicked displays the current status and balance of channels from the DB. They also display similar options to the above:
* send via channel (minidapp only) if the party's channel balance > 0
* request via channel (implemented correctly only in the android app)
* trigger settlement

The settlement options work the same way whether clicked in the view above (last screenshot of previous comment) or from the channel listing. Clicking "Trigger settlement" updates the channel status in the DB to "TRIGGERED" and posts the trigger transactions stored in the DB to Minima. The "Trigger settlement" button disappears. 
The eltoo script coins for each channel are checked upon every NEWBLOCK message from Minima. If found, their "created" block number + timelock value from the DB are compared with the current block number. Once elapsed, the "Complete settlement" button is displayed. 
The state value at port 99 is also checked at the eltoo script coins and if it's less the the sequence number stored in the DB, a "Post latest update" button appears. When clicked, the update transaction stored in the DB is posted to Minima and channel status in the DB updated. NB this will reset the timelock.
If the timelock elapsed and the "Complete settlement" button is clicked, the settlement transaction stored in the DB is posted to Minima.
TODO: a warning should be displayed if a channel is triggered and the counterparty posts an update transaction with a lower sequence number (state 99).

**Comment by mihbor** [comment]
Channel updates.
In the minidapp if you click the "Send via channel" button, a new update+settlement transaction pair is created debiting oneself and crediting the counterparty and with an incremented sequence number (state 99). They are signed exported and send over firebase with TXN_UPDATE prefix.
The receiver imports, signs and exports the transactions and updates the channel in their DB (notably the balances, exported transactions and sequence number) and sends them back over firebase with TXN_UPDATE_ACK prefix. Once the sender receives the ack, it updates it's channel in the DB (notably the balances, exported transactions, which are now signed by both parties and the sequence number) .
Firebase subscriptions for all channels stored in the DB with status = "OPEN" are re-established on application startup.

In the android app only the "Request via channel" option is implemented. Once clicked, a new update+settlement transaction pair is created debiting the counterparty and crediting oneself and with an incremented sequence number. They are signed, exported, prefixed with "TXN_REQUEST" and sent to the CardService, which will emit it over NFC on request. A dialogue is displayed informing the user to press the "Request sent" button once the NFC exchange is complete. This will put the app back into "Scan" mode. 
When the counterparty puts their phone in NFC range and their app is in Scan mode, it will receive the transactions, import them, calculate the balance change and display a dialogue to the user to accept or reject it. Clicking "Reject" simply dismisses the dialogue screen. Clicking "Accept" will sign the transactions, export them and send them prefixed with "TXN_UPDATE_ACK" to the CardService which will emit them on request. The channel state is updated in the DB. The screen informs the user to interact via NFC again only offers a "Finish" button, which puts the app back into "Scan" mode.
When the requesting party puts their phone back in NFC range, it receives and imports the update and settlement transactions, which are now signed by both parties. It updates the channel in its DB and brings back the default screen.
